src/analysis/normed/ring/seminorm.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2022 María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos-Fernández
+-/
+import analysis.seminorm
+
+/-!
+# Seminorms and norms on rings
+
+This file defines seminorms and norms on rings. These definitions are useful when one needs to
+consider multiple (semi)norms on a given ring.
+
+## Main declarations
+
+For a ring `R`:
+* `ring_seminorm`: A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes
+  nonnegative values, is subadditive and submultiplicative and such that `f (-x) = f x` for all
+  `x ∈ R`.
+* `ring_norm`: A seminorm `f` is a norm if `f x = 0` if and only if `x = 0`.
+
+## References
+
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+ring_seminorm, ring_norm
+-/
+
+noncomputable theory
+
+set_option old_structure_cmd true
+
+open_locale nnreal
+
+variables {R S : Type*} (x y : R) (r : ℝ)
+
+/-- A seminorm on a ring `R` is a function `f : R → ℝ` that preserves zero, takes nonnegative
+  values, is subadditive and submultiplicative and such that `f (-x) = f x` for all `x ∈ R`. -/
+structure ring_seminorm (R : Type*) [non_unital_ring R]
+  extends add_group_seminorm R :=
+(mul_le' : ∀ x y : R, to_fun (x * y) ≤ to_fun x * to_fun y)
+
+attribute [nolint doc_blame] ring_seminorm.to_add_group_seminorm
+
+namespace ring_seminorm
+
+section non_unital_ring
+
+variables [non_unital_ring R]
+
+instance zero_hom_class : zero_hom_class (ring_seminorm R) R ℝ :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_zero := λ f, f.map_zero' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun (ring_seminorm R) (λ _, R → ℝ) := ⟨λ p, p.to_fun⟩
+
+@[simp] lemma to_fun_eq_coe (p : ring_seminorm R) : p.to_fun = p := rfl
+
+@[ext] lemma ext {p q : ring_seminorm R} (h : ∀ x, p x = q x) : p = q :=
+fun_like.ext p q h
+
+instance : has_zero (ring_seminorm R) :=
+⟨{ mul_le' :=  λ _ _, eq.ge (zero_mul _),
+  ..add_group_seminorm.has_zero.zero }⟩
+
+lemma eq_zero_iff {p : ring_seminorm R} : p = 0 ↔ ∀ x, p x = 0 := fun_like.ext_iff
+lemma ne_zero_iff {p : ring_seminorm R} : p ≠ 0 ↔ ∃ x, p x ≠ 0 := by simp [eq_zero_iff]
+
+instance : inhabited (ring_seminorm R) := ⟨0⟩
+
+variables (p : ring_seminorm R)
+
+protected lemma nonneg : 0 ≤ p x := p.nonneg' _
+@[simp] protected lemma map_zero : p 0 = 0 := p.map_zero'
+protected lemma add_le : p (x + y) ≤ p x + p y := p.add_le' _ _
+@[simp] protected lemma neg : p (- x) = p x := p.neg' _
+protected lemma mul_le : p (x * y) ≤ p x * p y := p.mul_le' _ _
```

Comment:
You should not need all this anymore.

Reply:
I have removed all but `nonneg` and `mul_le`. I will get rid of those in the PR to introduce `ring_seminorm_class`/`ring_norm_class`.

Reply:
See https://github.com/leanprover-community/mathlib/pull/14115#discussion_r970476168